### PR TITLE
Port sound player and sound tokens from Bay

### PR DESCRIPTION
**Pull request by FluffyGhoster**
This PR ports the sound player and sound tokens from Bay and rewrites the jukebox to use it.

This is a continuation from #15519 (based on its branch) and will need to be merged after it to guarantee atomicity.

Implemented __datastructures\stack.dm as a dependency of them, eventually other datastructures should be moved to said folder for structure cleanliness.

### CODE REVIEW

**Comment by FluffyGhoster** [comment]
Ready for review, should be an easy one this time (eye of the cyclone however, probably)

**Review by alsoandanswer — APPROVED** [review]
Should name your changelogs as
[username]_[PR name]

File: code/datums/sound_player.dm [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	// Once stopped, always stopped. Go ask the player to play the sound again.
+	if(status & SOUND_STOPPED)
+		return
+	if(new_status == status)
+		return
+	status = new_status
+	PrivUpdateListeners()
+
+/datum/sound_token/proc/PrivAddListener(atom/listener)
+	// if(isvirtualmob(listener))
```
Comment (NonQueueingMatt) [inline code/datums/sound_player.dm:199]: commented code

File: code/datums/sound_player.dm [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		sound.channel = channel
+	else
+		sound.channel = 0
+
+	listeners = list()
+	listener_status = list()
+
+	destroyed_event.register(source, src, /datum/proc/qdel_self)
+
+	// if(ismovable(source))
```
Comment (NonQueueingMatt) [inline code/datums/sound_player.dm:119]: commented code

File: code/datums/sound_player.dm [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	if(islist(environment) && length(environment) != 23)
+		return FALSE
+	if(!isnum(environment) || environment < 0 || environment > 25)
+		return FALSE
+	return TRUE
+
+/datum/sound_token/static_environment/PrivGetEnvironment()
+	return sound.environment
+
+// /obj/sound_test
```
Comment (NonQueueingMatt) [inline code/datums/sound_player.dm:274]: commented code

File: code/game/machinery/jukebox.dm [hunk 4]
```diff
@@ -184,10 +185,12 @@ datum/track/New(var/title_name, var/audio)
 		return 1
 
 /obj/machinery/media/jukebox/proc/StopPlaying()
-	var/area/main_area = get_area(src)
-	// Always kill the current sound
-	for(var/mob/living/M in mobs_in_area(main_area))
-		M << sound(null, channel = 4)
+	// var/area/main_area = get_area(src)
```
Comment (NonQueueingMatt) [inline code/game/machinery/jukebox.dm:188]: commented code, shouldn't this be uncommented?

File: html/changelogs/fluffyghost.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+# Any changes you've made.  See valid prefix list above.
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, all entries are changed into a single [] after a master changelog generation. Just remove the brackets when you add new entries.
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes:
+  - refactor: "The jukebox now uses the sound player."
+  - rscadd: "Added Bays sound player and sound tokens."
+  - rscadd: "Added (DOUBLE UNDERSCORE)datastructures(BACKSLASH)stack data structure, used by the sound player and token."
```
Comment (NonQueueingMatt) [inline html/changelogs/fluffyghost.yml:43]: People don't need to know this, and it looks like a really ugly changelog entry

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]
These new sound loops should substitute some (or all) of the current ones that use TIMER_LOOP. Also, there's a lot of commented code that needs to be removed.

File: code/game/machinery/jukebox.dm (hunk 4 above)
Comment (FluffyGhoster) [reply code/game/machinery/jukebox.dm:188]: I don't think this should be uncommented, the tokens handles the stopping of the sound when they get QDELed, making it unnecessary to handle it in other parts

In the Destroy() function, it calls Stop() before going up the parent chain, Stop() takes care of pushing the null sound to the mobs to stop the sound, deregister from the SSProcessing etc

File: code/__datastructures/stack.dm [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/datum/stack
```
Comment (NonQueueingMatt) [inline code/__datastructures/stack.dm:1]: We already have this under `/Stack`. Remove one of the two.

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

File: code/__datastructures/stack.dm (hunk 6 above)
Comment (FluffyGhoster) [reply code/__datastructures/stack.dm:1]: We also had a third one from 11 years ago, removed that one too, neither were even used

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**